### PR TITLE
More BackendCompatible generalizations

### PR DESCRIPTION
**Pull request by parsonsmatt**
This PR includes some additional generalizations. Unfortunately, my work project does not use the `SqlReadT` or `SqlWriteT` features, and it was only when the code was tested on those did the issues with Esqueleto come up.

This PR generalizes `SqlReadT` so that Esqueleto can be updated in a way that doesn't cause any breaking changes.

If any folks are using SqlReadT, I'd appreciate testing this PR/commit out with [this Esqueleto PR](https://github.com/bitemyapp/esqueleto/pull/53) to verify that it doesn't break any corner cases I've missed :smile:

### CODE REVIEW

File: persistent/Database/Persist/Sql/Raw.hs [hunk 1]
```diff
@@ -47,20 +47,20 @@ rawQueryRes sql vals = do
         stmtQuery stmt vals
 
 -- | Execute a raw SQL statement
-rawExecute :: MonadIO m
+rawExecute :: (MonadIO m, BackendCompatible SqlBackend backend)
```
Comment (parsonsmatt) [inline persistent/Database/Persist/Sql/Raw.hs]: These functions needed to be generalized to satisfy Esqueleto.

File: persistent/Database/Persist/Sql/Migration.hs [hunk 2]
```diff
@@ -27,6 +27,7 @@ import Control.Monad.Trans.Control (liftBaseOp_)
 import Database.Persist.Sql.Types
 import Database.Persist.Sql.Raw
 import Database.Persist.Types
+import Database.Persist.Sql.Orphan.PersistStore()
```
Comment (parsonsmatt) [inline persistent/Database/Persist/Sql/Migration.hs]: This import is necessary for the `BackendCompatible SqlBackend SqlBackend` instance.

**Comment by dariooddenino** [comment]
Compilation of `persistent-mysql-2.6.2` fails with this branch. `persistent-mysql-2.6.1` works fine.

```
Progress: 2/3
--  While building package persistent-mysql-2.6.2 using:
      /home/dario/.stack/setup-exe-cache/x86_64-linux-nopie/Cabal-simple_mPHDZzAJ_1.24.2.0_ghc-8.0.2 --builddir=.stack-work/dist/x86_64-linux-nopie/Cabal-1.24.2.0 build --ghc-options " -ddump-hi -ddump-to-file"
    Process exited with code: ExitFailure 1
    Logs have been written to: /home/dario/dev/blue-moon/.stack-work/logs/persistent-mysql-2.6.2.log

    Configuring persistent-mysql-2.6.2...
    Building persistent-mysql-2.6.2...
    Preprocessing library persistent-mysql-2.6.2...
    [1 of 1] Compiling Database.Persist.MySQL ( Database/Persist/MySQL.hs, .stack-work/dist/x86_64-linux-nopie/Cabal-1.24.2.0/build/Database/Persist/MySQL.o )
    
    /tmp/stack4584/persistent-mysql-2.6.2/Database/Persist/MySQL.hs:1188:17: error:
        • Could not deduce (BackendCompatible r0 backend)
            arising from a use of ‘projectBackend’
          from the context: (backend ~ PersistEntityBackend record,
                             BackendCompatible SqlBackend backend,
                             PersistEntity record,
                             MonadIO m)
            bound by the type signature for:
                       insertManyOnDuplicateKeyUpdate :: (backend
                                                          ~
                                                          PersistEntityBackend record,
                                                          BackendCompatible SqlBackend backend,
                                                          PersistEntity record, MonadIO m) =>
                                                         [record]
                                                         -> [SomeField record]
                                                         -> [Update record]
                                                         -> ReaderT backend m ()
            at Database/Persist/MySQL.hs:(1175,1)-(1185,27)
          The type variable ‘r0’ is ambiguous
          Relevant bindings include
            insertManyOnDuplicateKeyUpdate :: [record]
                                              -> [SomeField record]
                                              -> [Update record]
                                              -> ReaderT backend m ()
              (bound at Database/Persist/MySQL.hs:1186:1)
          These potential instances exist:
            instance BackendCompatible SqlBackend SqlBackend
              -- Defined in ‘persistent-2.7.1:Database.Persist.Sql.Orphan.PersistStore’
            instance BackendCompatible SqlBackend SqlReadBackend
              -- Defined in ‘persistent-2.7.1:Database.Persist.Sql.Orphan.PersistStore’
            instance BackendCompatible SqlBackend SqlWriteBackend
              -- Defined in ‘persistent-2.7.1:Database.Persist.Sql.Orphan.PersistStore’
        • In the first argument of ‘withReaderT’, namely ‘projectBackend’
          In the first argument of ‘(.)’, namely ‘withReaderT projectBackend’
          In the expression: withReaderT projectBackend . uncurry rawExecute
    
    /tmp/stack4584/persistent-mysql-2.6.2/Database/Persist/MySQL.hs:1189:15: error:
        • Could not deduce (BackendCompatible SqlBackend r0)
            arising from a use of ‘rawExecute’
          from the context: (backend ~ PersistEntityBackend record,
                             BackendCompatible SqlBackend backend,
                             PersistEntity record,
                             MonadIO m)
            bound by the type signature for:
                       insertManyOnDuplicateKeyUpdate :: (backend
                                                          ~
                                                          PersistEntityBackend record,
                                                          BackendCompatible SqlBackend backend,
                                                          PersistEntity record, MonadIO m) =>
                                                         [record]
                                                         -> [SomeField record]
                                                         -> [Update record]
                                                         -> ReaderT backend m ()
            at Database/Persist/MySQL.hs:(1175,1)-(1185,27)
          The type variable ‘r0’ is ambiguous
          These potential instances exist:
            instance BackendCompatible SqlBackend SqlBackend
              -- Defined in ‘persistent-2.7.1:Database.Persist.Sql.Orphan.PersistStore’
            instance BackendCompatible SqlBackend SqlReadBackend
              -- Defined in ‘persistent-2.7.1:Database.Persist.Sql.Orphan.PersistStore’
            instance BackendCompatible SqlBackend SqlWriteBackend
              -- Defined in ‘persistent-2.7.1:Database.Persist.Sql.Orphan.PersistStore’
        • In the first argument of ‘uncurry’, namely ‘rawExecute’
          In the second argument of ‘(.)’, namely ‘uncurry rawExecute’
          In the expression: withReaderT projectBackend . uncurry rawExecute
```

**Comment by parsonsmatt** [comment]
Good catch @dariooddenino The generalizations I'm making here make the preemptive `persistent-mysql` generalizations unnecessary :)

**Comment by paul-rouse** [comment]
@parsonsmatt This is looking good!  Could you please rebase to the latest master, and make the changelog message describe the whole PR (include a link to the PR too) instead of saying it fixes a compilation error (which was just an intermediate step!)?  Then, assuming travis is still OK, I can merge it.

**Comment by parsonsmatt** [comment]
Weird spurious build failure: https://travis-ci.org/yesodweb/persistent/jobs/296983378 Tests passed and all :\

**Comment by paul-rouse** [comment]
Looked like a last-minute glitch. I restarted that test and it is fine now.

Thanks for all of this - I'll go ahead and merge it.

**Pull request merged**